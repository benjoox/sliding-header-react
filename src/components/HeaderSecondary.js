import React, { useState, useRef, useEffect } from 'react'
import { bool, number } from 'prop-types'
import DropdownItem from './DropdownItem'
import TopicItem from './TopicItem'

const menuList = [
  'Brexit',
  'Climate',
  'Copyright',
  'Migration',
  'Deb on hover',
  'Debate X',
  'Debate Y',
  'Debate L',
  'Debate S',
]
export default function HeaderSecondary({ width, collapsed }) {
  const [widthArray, setWidthArray] = useState([])
  // Number of items in the primary header
  const [primaryItemCount, setPrimaryItemCount] = useState(menuList.length)

  // Number of items in the secondary header
  const [secondaryItemCount, setSecondaryItemCount] = useState(menuList.length)
  const ulRef = useRef()

  useEffect(() => {
    const allowedWidth = width - 290 - 120
    function calculatePrimaryItems() {
      let sumWidth = 0
      for (let k = 0; k < widthArray.length; k++) {
        sumWidth = sumWidth + widthArray[k]
        if (sumWidth > width) {
          setSecondaryItemCount(k)
          break
        }
      }
    }

    function calculateSecondaryItems(allowedWidth) {
      let sumWidth = 0
      for (let k = 0; k < widthArray.length; k++) {
        sumWidth = sumWidth + widthArray[k]
        if (sumWidth > allowedWidth) {
          setPrimaryItemCount(k)
          break
        }
      }
    }
    if (widthArray.length === menuList.length && allowedWidth > 0) {
      calculatePrimaryItems()
      calculateSecondaryItems(allowedWidth)
    }
  }, [width, widthArray, collapsed])

  const topicList = () => {
    const result = []
    for (let k = 0; k < menuList.length; k++) {
      if (k < secondaryItemCount) {
        result.push(
          <TopicItem
            key={k}
            topic={menuList[k]}
            index={k}
            className={k > primaryItemCount && collapsed ? 'fade-out' : ''}
            setElementWidth={(index, width) => {
              widthArray[index] = width
              setWidthArray(widthArray)
            }}
          />
        )
      }
    }
    return result
  }

  return (
    <div className={`secondary ${collapsed ? 'collapsed' : ''}`}>
      <ul ref={ulRef}>
        <DropdownItem />
        {topicList()}
      </ul>
    </div>
  )
}

HeaderSecondary.propTypes = {
  width: number.isRequired,
  collapsed: bool.isRequired,
}
