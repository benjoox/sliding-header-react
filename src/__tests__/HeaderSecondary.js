import React from 'react'
import renderer from 'react-test-renderer'
import HeaderSecondary from '../components/HeaderSecondary'

it('Large screen - HeaderSecondary renders correctly when collpased', () => {
  const tree = renderer
    .create(<HeaderSecondary collapsed={true} width={1080} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Large screen - HeaderSecondary renders correctly when expanded', () => {
  const tree = renderer
    .create(<HeaderSecondary collapsed={false} width={1080} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Small screen - HeaderSecondary renders correctly when collpased', () => {
  const tree = renderer
    .create(<HeaderSecondary collapsed={true} width={575} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Small screen - HeaderSecondary renders correctly when expanded', () => {
  const tree = renderer
    .create(<HeaderSecondary collapsed={false} width={575} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
