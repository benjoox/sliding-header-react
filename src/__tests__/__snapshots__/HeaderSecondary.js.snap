// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Large screen - HeaderSecondary renders correctly when collpased 1`] = `
<div
  className="secondary collapsed"
>
  <ul>
    <li
      className="dropdown-item"
    >
      All A-Z
      <span>
        <svg>
          arrow-down.svg
        </svg>
      </span>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Brexit
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Climate
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Copyright
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Migration
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Deb on hover
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate X
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate Y
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate L
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate S
      </a>
    </li>
  </ul>
</div>
`;

exports[`Large screen - HeaderSecondary renders correctly when expanded 1`] = `
<div
  className="secondary "
>
  <ul>
    <li
      className="dropdown-item"
    >
      All A-Z
      <span>
        <svg>
          arrow-down.svg
        </svg>
      </span>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Brexit
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Climate
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Copyright
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Migration
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Deb on hover
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate X
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate Y
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate L
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate S
      </a>
    </li>
  </ul>
</div>
`;

exports[`Small screen - HeaderSecondary renders correctly when collpased 1`] = `
<div
  className="secondary collapsed"
>
  <ul>
    <li
      className="dropdown-item"
    >
      All A-Z
      <span>
        <svg>
          arrow-down.svg
        </svg>
      </span>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Brexit
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Climate
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Copyright
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Migration
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Deb on hover
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate X
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate Y
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate L
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate S
      </a>
    </li>
  </ul>
</div>
`;

exports[`Small screen - HeaderSecondary renders correctly when expanded 1`] = `
<div
  className="secondary "
>
  <ul>
    <li
      className="dropdown-item"
    >
      All A-Z
      <span>
        <svg>
          arrow-down.svg
        </svg>
      </span>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Brexit
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Climate
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Copyright
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Migration
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Deb on hover
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate X
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate Y
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate L
      </a>
    </li>
    <li
      className="topic-item"
    >
      <a
        href="#"
      >
        Debate S
      </a>
    </li>
  </ul>
</div>
`;
